File: spec/java/src/io/kaitai/struct/specwrite/TestEosExceptionU4.java
package io.kaitai.struct.specwrite;

import io.kaitai.struct.ByteBufferKaitaiStream;
import io.kaitai.struct.KaitaiStream;
import io.kaitai.struct.KaitaiStruct.ReadWrite;
import io.kaitai.struct.testwrite.EosExceptionU4;
import org.testng.annotations.Test;
import org.testng.SkipException;

public class TestEosExceptionU4 extends CommonSpec {
    @Override
    protected Class<? extends ReadWrite> getStructClass() {
        return EosExceptionU4.class;
    }

    @Override
    protected String getSrcFilename() {
        return "term_strz.bin";
    }

    @Override
    @Test
    protected void testReadWriteRoundtrip() throws Exception {
        throw new SkipException("cannot use roundtrip because parsing is expected to fail");
    }

    @Test(expectedExceptions = java.nio.BufferOverflowException.class)
    public void testEosExceptionU4() throws Exception {
        EosExceptionU4 r = new EosExceptionU4();

        EosExceptionU4.Data data = new EosExceptionU4.Data(null, r, r._root());
        data.setPrebuf(new byte[] { 0, -1, -2 });
        data.setFailInt(3000500200L);
        data._check();

        r.setEnvelope(data);
        r._check();

        try (KaitaiStream io = new ByteBufferKaitaiStream(12)) {
            r._write(io);
        }
    }
}

File: spec/python/specwrite/test_eos_exception_u4.py
import unittest
import io
from kaitaistruct import KaitaiStream
from specwrite.common_spec import CommonSpec

from testwrite.eos_exception_u4 import EosExceptionU4

class TestEosExceptionU4(CommonSpec.Base):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.struct_class = EosExceptionU4
        self.src_filename = 'src/term_strz.bin'

    def test_read_write_roundtrip(self):
        self.skipTest("cannot use roundtrip because parsing is expected to fail")

    def test_eos_exception_u4(self):
        r = EosExceptionU4()

        data = EosExceptionU4.Data(None, r, r._root)
        data.prebuf = b"\x00\xFF\xFE"
        data.fail_int = 3000500200
        data._check()

        r.envelope = data
        r._check()

        with KaitaiStream(io.BytesIO(bytes(12))) as out_io:
            with self.assertRaisesRegex(EOFError, "^requested to write 4 bytes, but only 3 bytes left in the stream$"):
                r._write(out_io)
